File: crates/ih-muse-proto/src/element_kind.rs
// crates/ih-muse-proto/src/element_kind.rs

use serde::{Deserialize, Serialize};

use crate::types::*;
use crate::utils::deterministic_u64_from_str;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ElementKindRegistration {
    id: ElementKindId,
    pub code: String,
    parent_id: Option<ElementKindId>,
    pub name: String,
    pub description: String,
    forwarded: bool,
}

impl ElementKindRegistration {
    pub fn new(
        code: String,
        parent_code: Option<String>,
        name: String,
        description: String,
    ) -> Self {
        Self {
            id: deterministic_u64_from_str(&code),
            code,
            parent_id: parent_code.as_deref().map(deterministic_u64_from_str),
            name,
            description,
            forwarded: false,
        }
    }
}

File: py-ih-muse/ih_muse/proto/element_kind.py
# py-ih-muse/ih_muse/muse/muse.py
from __future__ import annotations

from typing import Optional

from ih_muse.ih_muse import PyElementKindRegistration


class ElementKindRegistration:
    _elem_kind_reg: PyElementKindRegistration

    def __init__(
        self, code: str, name: str, description: str, parent_code: Optional[str] = None
    ) -> None:
        self._elem_kind_reg = PyElementKindRegistration(
            code, name, description, parent_code
        )
